thinc/loss.py

```diff
@@ -98,21 +97,38 @@ def convert_truths(self, truths, guesses: Floats2d) -> Tuple[Floats2d, Floats2d]
                     truths = [self._name_to_i[name] for name in truths]
             truths = xp.asarray(truths, dtype="i")
             mask = _make_mask(guesses, missing)
+        # Deal with truths in xp array format.
         else:
             mask = _make_mask_by_value(truths, guesses, missing_value)
-        if truths.ndim != guesses.ndim:
-            # transform categorical values to one-hot encoding
-            truths = to_categorical(
-                cast(Ints1d, truths),
-                n_classes=guesses.shape[-1],
-                label_smoothing=self.label_smoothing,
-            )
-        else:
-            if self.label_smoothing:
+            # Convert 1d truths to 2d and apply smoothing.
+            if truths.ndim == 1:
+                truths = to_categorical(
+                    cast(Ints1d, truths),
+                    n_classes=guesses.shape[-1],
+                    label_smoothing=self.label_smoothing,
+                )
+            # Validate 2d truths and apply smoothing if its one-hot.
+            elif truths.ndim == 2:
+                if not xp.all(truths.sum(axis=1) == 1):
+                    raise ValueError(
+                        "Cannot calcuate CategoricalCrossentropy. "
+                        "All rows of 'truths' have to be a "
+                        "valid categorical distribution (sum to 1)."
+                    )
+                if self.label_smoothing:
+                    # Check if one-hot
+                    if xp.all(truths.sum(axis=0) == 1):
+                        truths = smooth_one_hot(truths, self.label_smoothing)
+                    else:
+                        raise ValueError(
+                            "Can only apply label-smoothing to one-hot target."
+                        )
+            # Something went wrong.
+            else:
```

Comment:
Maybe with these changes, we can also specify the type in the signature for `truths`?

Reply:
okok good point!

Reply:
Ah right, I tried it before, its just really hard to get `mypy` to be happy, but I'll try!